### PR TITLE
Add pprof proto output as command line arg. to perf profiler test.

### PR DESCRIPTION
**Pull request by etep**
Summary: Adds pprof proto output to profiler test via a command line arg.
```
scripts/sudo_bazel_run.sh //src/stirling/source_connectors/perf_profiler:perf_profiler_bpf_test -- -pprof_pb_file cpp-test.pprof.pb --gtest_filter=PerfProfileBPFTest.PerfProfilerCppTest
```

Currently, we are using this to investigate flaky tests. We plan to (in a subsequent PR) refactor the pprof export code into a library function that enables pprof export from pixel scripts or from a standalone pem.

Relevant Issues: https://github.com/pixie-io/pixie/issues/719

Type of change: /kind feature

Test Plan: Created a pprof proto file for each profiler test.

### CODE REVIEW

File: src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc [hunk 1]
```diff
@@ -251,6 +259,143 @@ class PerfProfileBPFTest : public ::testing::TestWithParam<std::filesystem::path
...
+
+    // Create a file.
+    std::ofstream ofs(FLAGS_pprof_pb_file, std::ios_base::out | std::ios_base::binary);
+
+    if (!ofs.is_open()) {
+      return error::ResourceUnavailable("Could not create or open file: $0.", FLAGS_pprof_pb_file);
+    }
+
+    // Serialize the pprof proto to that file.
+    if (!profile.SerializeToOstream(&ofs)) {
```
Comment (ddelnano) [inline src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc:384]: `WritePProfProtoFile` seems to be called by multiple tests. My understanding is that the resulting pprof file ends up being whichever test ran last. Is that correct?

Comment (etep) [reply src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc:384]: Yes, that's true. In the example usage, I show the test case with a gtest filter (i.e. so that only one test is run and then the resulting pprof proto is unambiguous). I will figure out a way to either clarify the usage, or fail if the feature is abused, or uniquify the file name.

File: bazel/repository_locations.bzl [hunk 2]
```diff
@@ -459,6 +459,13 @@ GIT_REPOSITORY_LOCATIONS = dict(
         commit = "fa36a4fa141b4e9486b9126640d54a94c1d36fce",
         shallow_since = "1638898188 -0800",
     ),
+
+    # Used by the perf profiler to create pprof proto output.
+    com_github_google_pprof = dict(
```
Comment (JamesMBartlett) [inline bazel/repository_locations.bzl:464]: Please use http repos not git repos. git repos don't use the repository cache

File: src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc [hunk 3]
```diff
@@ -251,6 +259,143 @@ class PerfProfileBPFTest : public ::testing::TestWithParam<std::filesystem::path
     VLOG(1) << std::endl;
   }
 
+  Status WritePProfProtoFile() {
```
Comment (JamesMBartlett) [inline src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc:262]: Im a little confused about the purpose of this in the test. It seems unrelated to testing the per profiler functionality. Is it just for debugging the test?

File: src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc (hunk 1 above)
Comment (ddelnano) [reply src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc:384]: What if we specified the file prefix and then suffixed test name or test index on the end of it? I think naming the flag that way will make the usage more obvious as well.

File: bazel/repository_locations.bzl (hunk 2 above)
Comment (etep) [reply bazel/repository_locations.bzl:464]: Hi @JamesMBartlett,

Could you take a look at the upstream pprof repo on github? I did not see any tags or any other artifacts that looked compatible with `http_archive` or `_bazel_repo`. Maybe there is a way to make it work, but it was not obvious to me. (or maybe you are suggesting a different way to bring the repo in). I had left a comment (above) where I invoked `_git_repo`:
```
# Using "git repo" because the pprof repo does not have any release tags.
```

File: src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc (hunk 3 above)
Comment (etep) [reply src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc:262]: This is what I had in the summary of the PR:
```
Currently, we are using this to investigate flaky tests. We plan to (in a subsequent PR) refactor the pprof export code into a library function that enables pprof export from pixel scripts or from a standalone pem.
```

Since we are using this for test debug, I figured it was a nice moment to just review the basic pprof export code. We may refactor it into a library function soon.

Let's touch base in a different forum if you want to check out the results of the flaky test investigation.

**Review by zasgar — CHANGES_REQUESTED** [review]

File: bazel/repository_locations.bzl (hunk 2 above)
Comment (JamesMBartlett) [reply bazel/repository_locations.bzl:464]: We do this in a number of places in this file, you can just point to the commit itself. It's not ideal because github changes their hashes sometimes, but that's a separate issue we need to figure out. 
eg: "https://github.com/antlr/grammars-v4/archive/e53d7a1228505bfc80d8637808ef60e7eea92cc2.tar.gz"

File: src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc [hunk 4]
```diff
@@ -239,6 +242,11 @@ class PerfProfileBPFTest : public ::testing::TestWithParam<std::filesystem::path
       histo_[stack_trace_str] += count;
     }
```
[the remark this answers is not in the document]
Comment (zasgar) [reply src/stirling/source_connectors/perf_profiler/perf_profiler_bpf_test.cc:244]: I think it's better for us to make it a library first. If we want to write out files, it's probably best to write it in a utility exec. 

For verbose logging, we probably should not be writing to a file in the test.
